src/parser/parser.rs

```diff
@@ -0,0 +1,435 @@
+            "==" => Expression::EQ(Box::new(left), Box::new(right)),
+            _ => unreachable!(),
+        },
+    ))
+}
+
+// Parse expressions with operator precedence
+fn arithmetic_expression(input: &str) -> IResult<&str, Expression> {
+    let (input, first) = term(input)?;
+    let (input, operations) = many0(tuple((delimited(space0, operator, space0), term)))(input)?;
+
+    Ok((
+        input,
+        operations
+            .into_iter()
+            .fold(first, |acc, (op, val)| match op {
+                "+" => Expression::Add(Box::new(acc), Box::new(val)),
+                "-" => Expression::Sub(Box::new(acc), Box::new(val)),
+                "*" => Expression::Mul(Box::new(acc), Box::new(val)),
+                "/" => Expression::Div(Box::new(acc), Box::new(val)),
+                _ => unreachable!(),
+            }),
+    ))
+}
+//indented block parser
+fn indented_block(input: &str) -> IResult<&str, Vec<Statement>> {
+    let (input, _) = line_ending(input)?;
+    let (input, statements) = separated_list1(
+        line_ending,
+        preceded(
+            space1, // Require at least one space for indentation
+            statement,
+        ),
+    )(input)?;
+    Ok((input, statements))
+}
+
+fn if_statement(input: &str) -> IResult<&str, Statement> {
+    let (input, _) = tag("if")(input)?;
+    let (input, _) = space1(input)?;
+    let (input, condition) = comparison_expression(input)?;
+    let (input, _) = space0(input)?;
+    let (input, _) = char(':')(input)?;
+    let (input, then_block) = indented_block(input)?;
+
+    let (input, else_block) = opt(preceded(
+        tuple((line_ending, space0, tag("else"), char(':'))),
+        indented_block,
+    ))(input)?;
+
+    Ok((
+        input,
+        Statement::IfThenElse(
+            Box::new(condition),
+            Box::new(Statement::Block(then_block)),
+            else_block.map(|stmts| Box::new(Statement::Block(stmts))), // Changed this line
+        ),
+    ))
+}
+fn declaration(input: &str) -> IResult<&str, Statement> {
+    let (input, keyword) = alt((tag("var"), tag("val")))(input)?;
+    let (input, _) = space1(input)?;
+    let (input, name) = identifier(input)?;
+
+    Ok((
+        input,
+        match keyword {
+            "var" => Statement::VarDeclaration(name), // No Box needed
+            "val" => Statement::ValDeclaration(name), // No Box needed
+            _ => unreachable!(),
+        },
+    ))
+}
+
+// Parse assignment statements
+fn assignment(input: &str) -> IResult<&str, Statement> {
+    let (input, _) = space0(input)?;
+    let (input, name) = identifier(input)?;
+    let (input, _) = delimited(space0, char('='), space0)(input)?;
+    let (input, expr) = expression(input)?;
+    let (input, _) = space0(input)?;
+
+    Ok((
+        input,
+        Statement::Assignment(
+            name, // No longer need to Box the name
+            Box::new(expr),
+        ),
+    ))
+}
+// Parse multiple statements
```

Comment:
minor, but adds up: keep the amount of redundant comments low. the function names pretty much describe what they do in this case. as we change/extend the functions, we may end up with obsolete comments